File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/glib/models/impl/LGroupArrowable.scala
package net.sf.latexdraw.glib.models.impl

import java.util.ArrayList
import java.util.List
import net.sf.latexdraw.glib.models.interfaces.shape.IArrow
import net.sf.latexdraw.glib.models.interfaces.shape.ArrowStyle
import net.sf.latexdraw.glib.models.interfaces.shape.IArrowableShape
import net.sf.latexdraw.glib.models.interfaces.shape.IGroup
import java.util.stream.Collectors

/**
 * This trait encapsulates the code of the group related to the support of arrowable shapes.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2014 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 2012-04-16<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
private[impl] trait LGroupArrowable extends IGroup {
	/** May return the first grid of the group. */
	private def firstIArrowable = arrowShapes.stream.filter{_.isTypeOf(classOf[IArrowableShape])}.findFirst

	private def arrowShapes = getShapes.stream.filter{_.isInstanceOf[IArrowableShape]}.map[IArrowableShape]{_.asInstanceOf[IArrowableShape]}.collect(Collectors.toList())


	override def getArrowIndex(arrow:IArrow) = if(firstIArrowable.isPresent) firstIArrowable.get.getArrowIndex(arrow) else -1

	override def getNbArrows = if(firstIArrowable.isPresent) firstIArrowable.get.getNbArrows else 0

	override def setTBarSizeDimList(values:List[java.lang.Double]) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setTBarSizeDim(values.get(j))
          case _ =>
        }
	}

	override def getTBarSizeDimList: List[java.lang.Double] = {
		val list = new ArrayList[java.lang.Double]()
		getShapes.forEach{
				case arr:IArrowableShape => list.add(arr.getTBarSizeDim)
				case _ => list.add(java.lang.Double.NaN)
		}
		return list
	}

	override def setTBarSizeNumList(values:java.util.List[java.lang.Double]) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setTBarSizeNum(values.get(j))
          case _ =>
        }
	}

	override def getTBarSizeNumList: java.util.List[java.lang.Double] = {
		val list = new ArrayList[java.lang.Double]()
		getShapes.forEach{
				case arr:IArrowableShape => list.add(arr.getTBarSizeNum)
				case _ => list.add(java.lang.Double.NaN)
		}
		return list
	}

	override def setDotSizeNumList(values:java.util.List[java.lang.Double]) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setDotSizeNum(values.get(j))
          case _ =>
        }
	}

	override def getDotSizeNumList: java.util.List[java.lang.Double] = {
		val list = new ArrayList[java.lang.Double]()
		getShapes.forEach{
				case arr:IArrowableShape => list.add(arr.getDotSizeNum)
				case _ => list.add(java.lang.Double.NaN)
		}
		return list
	}

	override def setDotSizeDimList(values:java.util.List[java.lang.Double]) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setDotSizeDim(values.get(j))
          case _ =>
        }
	}

	override def getDotSizeDimList: java.util.List[java.lang.Double] = {
		val list = new ArrayList[java.lang.Double]()
		getShapes.forEach{
				case arr:IArrowableShape => list.add(arr.getDotSizeDim)
				case _ => list.add(java.lang.Double.NaN)
		}
		return list
	}

	override def setBracketNumList(values:java.util.List[java.lang.Double]) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setBracketNum(values.get(j))
          case _ =>
        }
	}

	override def getBracketNumList: java.util.List[java.lang.Double] = {
		val list = new ArrayList[java.lang.Double]()
		getShapes.forEach{
				case arr:IArrowableShape => list.add(arr.getBracketNum)
				case _ => list.add(java.lang.Double.NaN)
		}
		return list
	}

	override def setRBracketNumList(values:java.util.List[java.lang.Double]) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setRBracketNum(values.get(j))
          case _ =>
        }
	}

	override def getRBracketNumList: java.util.List[java.lang.Double] = {
		val list = new ArrayList[java.lang.Double]()
		getShapes.forEach{
				case arr:IArrowableShape => list.add(arr.getRBracketNum)
				case _ => list.add(java.lang.Double.NaN)
		}
		return list
	}

	override def setArrowSizeNumList(values:java.util.List[java.lang.Double]) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setArrowSizeNum(values.get(j))
          case _ =>
        }
	}

	override def getArrowSizeNumList: java.util.List[java.lang.Double] = {
		val list = new ArrayList[java.lang.Double]()
		getShapes.forEach{
				case arr:IArrowableShape => list.add(arr.getArrowSizeNum)
				case _ => list.add(java.lang.Double.NaN)
		}
		return list
	}

	override def setArrowSizeDimList(values:java.util.List[java.lang.Double]) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setArrowSizeDim(values.get(j))
          case _ =>
        }
	}

	override def getArrowSizeDimList: java.util.List[java.lang.Double] = {
		val list = new ArrayList[java.lang.Double]()
		getShapes.forEach{
				case arr:IArrowableShape => list.add(arr.getArrowSizeDim)
				case _ => list.add(java.lang.Double.NaN)
		}
		return list
	}

	override def setArrowLengthList(values:java.util.List[java.lang.Double]) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setArrowLength(values.get(j))
          case _ =>
        }
	}

	override def getArrowLengthList: java.util.List[java.lang.Double] = {
		val list = new ArrayList[java.lang.Double]()
		getShapes.forEach{
				case arr:IArrowableShape => list.add(arr.getArrowLength)
				case _ => list.add(java.lang.Double.NaN)
		}
		return list
	}

	override def setArrowInsetList(values:java.util.List[java.lang.Double]) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setArrowInset(values.get(j))
          case _ =>
        }
	}

	override def getArrowInsetList: List[java.lang.Double] = {
		val list = new ArrayList[java.lang.Double]()
		getShapes.forEach{
				case arr:IArrowableShape => list.add(arr.getArrowInset)
				case _ => list.add(java.lang.Double.NaN)
		}
		return list
	}

	override def getArrowStyleList(i : Int) : java.util.List[ArrowStyle] =
		getShapes.stream.map[ArrowStyle]{case arr:IArrowableShape => arr.getArrowStyle(i); case _ => ArrowStyle.NONE}.collect(Collectors.toList())

	override def setArrowStyleList(values : java.util.List[ArrowStyle], i : Int) {
		val shapes = getShapes
		if(values!=null && values.size==shapes.size)
			for(j <- 0 until values.size)
				shapes.get(j) match {
          case shape: IArrowableShape => shape.setArrowStyle(values.get(j), i)
          case _ =>
        }
	}

	override def setArrowStyle(style : ArrowStyle, position : Int) {
		arrowShapes.forEach{_.setArrowStyle(style, position)}
	}

	override def getArrowStyle(position : Int) = if(firstIArrowable.isPresent) firstIArrowable.get.getArrowStyle(position) else ArrowStyle.NONE

	override def getArrowAt(position : Int) = if(firstIArrowable.isPresent) firstIArrowable.get.getArrowAt(position) else null

	override def getArrowLine(arrow : IArrow) = if(firstIArrowable.isPresent) firstIArrowable.get.getArrowLine(arrow) else null

	override def setDotSizeDim(dotSizeDim : Double) {
		arrowShapes.forEach{_.setDotSizeDim(dotSizeDim)}
	}

	override def setDotSizeNum(dotSizeNum : Double) {
		arrowShapes.forEach{_.setDotSizeNum(dotSizeNum)}
	}

	override def setTBarSizeNum(tbarSizeNum : Double) {
		arrowShapes.forEach{_.setTBarSizeNum(tbarSizeNum)}
	}

	override def setTBarSizeDim(tbarSizeDim : Double) {
		arrowShapes.forEach{_.setTBarSizeDim(tbarSizeDim)}
	}

	override def getTBarSizeDim = if(firstIArrowable.isPresent) firstIArrowable.get.getTBarSizeDim else Double.NaN

	override def getTBarSizeNum = if(firstIArrowable.isPresent) firstIArrowable.get.getTBarSizeNum else Double.NaN

	override def setRBracketNum(rBracketNum : Double) {
		arrowShapes.forEach{_.setRBracketNum(rBracketNum)}
	}

	override def setBracketNum(bracketNum : Double) {
		arrowShapes.forEach{_.setBracketNum(bracketNum)}
	}

	override def setArrowLength(lgth : Double) {
		arrowShapes.forEach{_.setArrowLength(lgth)}
	}

	override def setArrowSizeDim(arrowSizeDim : Double) {
		arrowShapes.forEach{_.setArrowSizeDim(arrowSizeDim)}
	}

	override def setArrowSizeNum(arrowSizeNum : Double) {
		arrowShapes.forEach{_.setArrowSizeNum(arrowSizeNum)}
	}

	override def setArrowInset(inset : Double) {
		arrowShapes.forEach{_.setArrowInset(inset)}
	}

	override def getDotSizeDim = if(firstIArrowable.isPresent) firstIArrowable.get.getDotSizeDim else Double.NaN

	override def getDotSizeNum = if(firstIArrowable.isPresent) firstIArrowable.get.getDotSizeNum else Double.NaN

	override def getBracketNum  = if(firstIArrowable.isPresent) firstIArrowable.get.getBracketNum else Double.NaN

	override def getArrowSizeNum = if(firstIArrowable.isPresent) firstIArrowable.get.getArrowSizeNum else Double.NaN

	override def getArrowSizeDim = if(firstIArrowable.isPresent) firstIArrowable.get.getArrowSizeDim else Double.NaN

	override def getArrowInset = if(firstIArrowable.isPresent) firstIArrowable.get.getArrowInset else Double.NaN

	override def getArrowLength = if(firstIArrowable.isPresent) firstIArrowable.get.getArrowLength else Double.NaN

	override def getRBracketNum = if(firstIArrowable.isPresent) firstIArrowable.get.getRBracketNum else Double.NaN
}

File: src/main/java/net/sf/latexdraw/models/impl/LGroupArrowable.java
/*
 * This file is part of LaTeXDraw.
 * Copyright (c) 2005-2018 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package net.sf.latexdraw.models.impl;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import net.sf.latexdraw.models.interfaces.shape.ArrowStyle;
import net.sf.latexdraw.models.interfaces.shape.IArrow;
import net.sf.latexdraw.models.interfaces.shape.IArrowableSingleShape;
import net.sf.latexdraw.models.interfaces.shape.IGroup;

/**
 * This trait encapsulates the code of the group related to the support of arrowable getShapes().
 * @author Arnaud Blouin
 */
interface LGroupArrowable extends IGroup {
	/** May return the first grid of the group. */
	default Optional<IArrowableSingleShape> firstIArrowable() {
		return arrowShapes().stream().filter(sh -> sh.isTypeOf(IArrowableSingleShape.class)).findFirst();
	}

	default List<IArrowableSingleShape> arrowShapes() {
		return getShapes().stream().filter(sh -> sh instanceof IArrowableSingleShape).map(sh -> (IArrowableSingleShape) sh).
			collect(Collectors.toList());
	}

	@Override
	default int getArrowIndex(final IArrow arrow) {
		return firstIArrowable().
			map(sh -> sh.getArrowIndex(arrow)).orElse(-1);
	}

	@Override
	default int getNbArrows() {
		return firstIArrowable().
			map(arc -> arc.getNbArrows()).orElse(0);
	}

	@Override
	default void setTBarSizeDimList(final List<Optional<Double>> values) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setTBarSizeDim(values.get(i).get()));
		}
	}

	@Override
	default List<Optional<Double>> getTBarSizeDimList() {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getTBarSizeDim()) : Optional.<Double>empty()).
			collect(Collectors.toList());
	}

	@Override
	default void setTBarSizeNumList(final List<Optional<Double>> values) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setTBarSizeNum(values.get(i).get()));
		}
	}

	@Override
	default List<Optional<Double>> getTBarSizeNumList() {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getTBarSizeNum()) : Optional.<Double>empty()).
			collect(Collectors.toList());
	}

	@Override
	default void setDotSizeNumList(final List<Optional<Double>> values) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setDotSizeNum(values.get(i).get()));
		}
	}

	@Override
	default List<Optional<Double>> getDotSizeNumList() {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getDotSizeNum()) : Optional.<Double>empty()).
			collect(Collectors.toList());
	}

	@Override
	default void setDotSizeDimList(final List<Optional<Double>> values) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setDotSizeDim(values.get(i).get()));
		}
	}

	@Override
	default List<Optional<Double>> getDotSizeDimList() {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getDotSizeDim()) : Optional.<Double>empty()).
			collect(Collectors.toList());
	}

	@Override
	default void setBracketNumList(final List<Optional<Double>> values) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setBracketNum(values.get(i).get()));
		}
	}

	@Override
	default List<Optional<Double>> getBracketNumList() {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getBracketNum()) : Optional.<Double>empty()).
			collect(Collectors.toList());
	}

	@Override
	default void setRBracketNumList(final List<Optional<Double>> values) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setRBracketNum(values.get(i).get()));
		}
	}

	@Override
	default List<Optional<Double>> getRBracketNumList() {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getRBracketNum()) : Optional.<Double>empty()).
			collect(Collectors.toList());
	}

	@Override
	default void setArrowSizeNumList(final List<Optional<Double>> values) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setArrowSizeNum(values.get(i).get()));
		}
	}

	@Override
	default List<Optional<Double>> getArrowSizeNumList() {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getArrowSizeNum()) : Optional.<Double>empty()).
			collect(Collectors.toList());
	}

	@Override
	default void setArrowSizeDimList(final List<Optional<Double>> values) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setArrowSizeDim(values.get(i).get()));
		}
	}

	@Override
	default List<Optional<Double>> getArrowSizeDimList() {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getArrowSizeDim()) : Optional.<Double>empty()).
			collect(Collectors.toList());
	}

	@Override
	default void setArrowLengthList(final List<Optional<Double>> values) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setArrowLength(values.get(i).get()));
		}
	}

	@Override
	default List<Optional<Double>> getArrowLengthList() {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getArrowLength()) : Optional.<Double>empty()).
			collect(Collectors.toList());
	}

	@Override
	default void setArrowInsetList(final List<Optional<Double>> values) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setArrowInset(values.get(i).get()));
		}
	}

	@Override
	default List<Optional<Double>> getArrowInsetList() {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getArrowInset()) : Optional.<Double>empty()).
			collect(Collectors.toList());
	}

	@Override
	default List<Optional<ArrowStyle>> getArrowStyleList(final int i) {
		return getShapes().stream().
			map(sh -> sh instanceof IArrowableSingleShape ? Optional.of(((IArrowableSingleShape) sh).getArrowStyle(i)) : Optional.<ArrowStyle>empty()).
			collect(Collectors.toList());
	}

	@Override
	default void setArrowStyleList(final List<Optional<ArrowStyle>> values, final int index) {
		if(values != null && values.size() == getShapes().size()) {
			IntStream.range(0, values.size()).
				filter(i -> getShapes().get(i) instanceof IArrowableSingleShape).
				forEach(i -> ((IArrowableSingleShape) getShapes().get(i)).setArrowStyle(values.get(i).get(), index));
		}
	}

	@Override
	default void setArrowStyle(final ArrowStyle style, final int position) {
		arrowShapes().
				forEach(sh -> sh.setArrowStyle(style, position));
	}

	@Override
	default ArrowStyle getArrowStyle(final int position) {
		return firstIArrowable().
			map(sh -> sh.getArrowStyle(position)).orElse(ArrowStyle.NONE);
	}

	@Override
	default IArrow getArrowAt(final int position) {
		return firstIArrowable().
			map(sh -> sh.getArrowAt(position)).orElse(null);
	}

	@Override
	default void setDotSizeDim(final double dotSizeDim) {
		arrowShapes().
				forEach(sh -> sh.setDotSizeDim(dotSizeDim));
	}

	@Override
	default void setDotSizeNum(final double dotSizeNum) {
		arrowShapes().
				forEach(sh -> sh.setDotSizeNum(dotSizeNum));
	}

	@Override
	default void setTBarSizeNum(final double tbarSizeNum) {
		arrowShapes().
				forEach(sh -> sh.setTBarSizeNum(tbarSizeNum));
	}

	@Override
	default void setTBarSizeDim(final double tbarSizeDim) {
		arrowShapes().
				forEach(sh -> sh.setTBarSizeDim(tbarSizeDim));
	}

	@Override
	default double getTBarSizeDim() {
		return firstIArrowable().
			map(sh -> sh.getTBarSizeDim()).orElse(Double.NaN);
	}

	@Override
	default double getTBarSizeNum() {
		return firstIArrowable().
			map(sh -> sh.getTBarSizeNum()).orElse(Double.NaN);
	}

	@Override
	default void setRBracketNum(final double rBracketNum) {
		arrowShapes().
				forEach(sh -> sh.setRBracketNum(rBracketNum));
	}

	@Override
	default void setBracketNum(final double bracketNum) {
		arrowShapes().
				forEach(sh -> sh.setBracketNum(bracketNum));
	}

	@Override
	default void setArrowLength(final double lgth) {
		arrowShapes().
				forEach(sh -> sh.setArrowLength(lgth));
	}

	@Override
	default void setArrowSizeDim(final double arrowSizeDim) {
		arrowShapes().
				forEach(sh -> sh.setArrowSizeDim(arrowSizeDim));
	}

	@Override
	default void setArrowSizeNum(final double arrowSizeNum) {
		arrowShapes().
				forEach(sh -> sh.setArrowSizeNum(arrowSizeNum));
	}

	@Override
	default void setArrowInset(final double inset) {
		arrowShapes().
				forEach(sh -> sh.setArrowInset(inset));
	}

	@Override
	default double getDotSizeDim() {
		return firstIArrowable().
			map(sh -> sh.getDotSizeDim()).orElse(Double.NaN);
	}

	@Override
	default double getDotSizeNum() {
		return firstIArrowable().
			map(sh -> sh.getDotSizeNum()).orElse(Double.NaN);
	}

	@Override
	default double getBracketNum() {
		return firstIArrowable().
			map(sh -> sh.getBracketNum()).orElse(Double.NaN);
	}

	@Override
	default double getArrowSizeNum() {
		return firstIArrowable().
			map(sh -> sh.getArrowSizeNum()).orElse(Double.NaN);
	}

	@Override
	default double getArrowSizeDim() {
		return firstIArrowable().
			map(sh -> sh.getArrowSizeDim()).orElse(Double.NaN);
	}

	@Override
	default double getArrowInset() {
		return firstIArrowable().
			map(sh -> sh.getArrowInset()).orElse(Double.NaN);
	}

	@Override
	default double getArrowLength() {
		return firstIArrowable().
			map(sh -> sh.getArrowLength()).orElse(Double.NaN);
	}

	@Override
	default double getRBracketNum() {
		return firstIArrowable().
			map(sh -> sh.getRBracketNum()).orElse(Double.NaN);
	}

	@Override
	default void setOnArrowChanged(final Runnable run) {
	}
}
